apps/src/music/blockly/blockUtils.js

```diff
@@ -56,3 +62,57 @@ export const isBlockInsideWhenRun = block => {
     block.getRootBlock().type
   );
 };
+
+// Override default function block implementation for the current block mode.
+export function installFunctionBlocks(blockMode) {
+  if (blockMode === BlockMode.ADVANCED) {
+    Blockly.cdoUtils.registerCustomAdvancedProcedureBlocks();
+    // Re-define blocks from core, in case they were deleted for Simple2 mode.
+    restoreBlockDefinitions();
+    // Replaces "variable:" with "parameter:" for added parameters
+    Blockly.Msg['PROCEDURE_VARIABLE'] = musicI18n.parameterLabel();
+  } else {
+    Blockly.cdoUtils.registerCustomProcedureBlocks();
+    // Remove two advanced blocks in the toolbox's Functions category that
+    // we don't want.
+    delete Blockly.Blocks.procedures_defreturn;
+    delete Blockly.Blocks.procedures_ifreturn;
+  }
+  // Sets the help URL for each function definiton block to the appropriate
+  // entry in the Music Lab docs.
+  Blockly.Msg['PROCEDURES_DEFRETURN_HELPURL'] =
+    DOCS_BASE_URL + 'create_function';
+  Blockly.Msg['PROCEDURES_DEFNORETURN_HELPURL'] =
+    DOCS_BASE_URL + 'create_function';
+}
+
+// Creates shallow copies of block definitions from core Blockly.
+// These definitions and overwritten by Simple2 but needed for advanced mode.
+// This makes it possible for us to switch block modes without a page reload.
+// See also: installFunctionBlocks
+export function backupFunctionDefinitons() {
+  Blockly.blockDefinitions = {};
+  const backupBlockDefinitionTypes = [
+    // Can potentially be overwritten by Simple2
+    BLOCK_TYPES.procedureCall,
+    // Can potentially be deleted by Simple2
+    BLOCK_TYPES.procedureIfReturn,
+  ];
+  backupBlockDefinitionTypes.forEach(type => {
+    Blockly.blockDefinitions[type] = Object.assign({}, Blockly.Blocks[type]);
```

Comment:
Should we be storing these on the `Blockly` global? Or should just be a music lab specific static object? Seems like we're only using this to cache the two procedure blocks right?

Reply:
Correct, and I had the same question myself. This logic is in a util, and it's called by our setup file, which also isn't a component. Is there a good place to cache these so that they'd be available when any of our components might need to handle an update?

Reply:
Maybe as a static member of `MusicBlocklyWorkspace`? We can even call `installFunctionBlocks` as a static method on `MusicBlocklyWorkspace` so its colocated with `setupBlocklyEnvironment()`

Reply:
In this commit, I just defined it as a constant in the util file. 

https://github.com/code-dot-org/code-dot-org/pull/61158/commits/4668996e755793d00b7a2f2c1ab81728936df669#diff-3fb497d1cec8fbfd63181c351d90a0de7e5f191cf445009d5d0761b388f0cf18

I can keep iterating if we don't like this though.

Reply:
Yeah that seems fine too - I was thinking it might be slightly nicer to keep everything routed through `MusicBlocklyWorkspace` but not a strong preference if this is easier to work with.